openmdao/solvers/solver_base.py

```diff
@@ -1,9 +1,75 @@
 """ Base class for linear and nonlinear solvers."""
 
 from __future__ import print_function
+
+from functools import wraps
+import sys
+from six import iterkeys, PY3
+
+import numpy as np
+
 from openmdao.recorders.recording_manager import RecordingManager
 from openmdao.util.options import OptionsDictionary
 
+def error_wrap_nl(fn):
+    """ Decorator adds some error-handling for floating point errors to any
+    driver function."""
+
+    @wraps(fn)
+    def wrapper(driver, params, unknowns, resids, system, metadata):
+        """ Mainly so sphinx autodoc works"""
+
+        try:
+            fn(driver, params, unknowns, resids, system, metadata)
+        except FloatingPointError as err:
+            exc_info = sys.exc_info()
+
+            # So we don't keep re-appending in a solver stack.
+            if hasattr(exc_info[1], 'seen'):
+                raise exc_info[0], exc_info[1], exc_info[2]
```

Comment:
need to use exec() here like you did below since this is a syntax error in python 3